src/ray/gcs/store_client/redis_store_client.h

```diff
@@ -0,0 +1,135 @@
+                  const StatusCallback &callback) override;
+
+  Status AsyncGet(const std::string &table_name, const std::string &key,
+                  const OptionalItemCallback<std::string> &callback) override;
+
+  Status AsyncGetByIndex(const std::string &table_name, const std::string &index,
+                         const MultiItemCallback<std::string> &callback) override;
+
+  Status AsyncGetAll(const std::string &table_name,
+                     const ScanCallback<std::string, std::string> &callback) override;
+
+  Status AsyncDelete(const std::string &table_name, const std::string &key,
+                     const StatusCallback &callback) override;
+
+  Status AsyncDeleteByIndex(const std::string &table_name, const std::string &index,
+                            const StatusCallback &callback) override;
+
+ private:
+  Status DoPut(const std::string &key, const std::string &value,
+               const StatusCallback &callback);
+
+  std::shared_ptr<RedisClient> redis_client_;
+};
+
+/// \class RedisRangeOpExecutor
+/// This class is used for three kind of operations:
+/// 1. Get All Data with the same prefix from Redis (GetAll).
+/// 2. Get All Data Key with the same index from Redis (GetByIndex).
+/// 3. Delete Index and Data Key with the same index from Redis (DeleteByIndex).
+/// TODO(micafan) Consider encapsulating three different classes:
+/// Scanner, MultiReader, BatchDeleter.
+class RedisRangeOpExecutor : public std::enable_shared_from_this<RedisRangeOpExecutor> {
+ public:
+  RedisRangeOpExecutor(std::shared_ptr<RedisClient> redis_client,
+                       const std::string &table_name, const std::string &index,
+                       const MultiItemCallback<std::string> &get_by_index_callback);
+
+  RedisRangeOpExecutor(std::shared_ptr<RedisClient> redis_client,
+                       const std::string &table_name, const std::string &index,
+                       const StatusCallback &delete_by_index_callback);
+
+  RedisRangeOpExecutor(std::shared_ptr<RedisClient> redis_client,
+                       const std::string &table_name,
+                       const ScanCallback<std::string, std::string> &get_all_callback);
+
+  ~RedisRangeOpExecutor();
+
+  Status Run();
+
+ private:
+  void OnFailed();
+
+  void OnDone();
+
+  void DoScan();
+
+  void DoCallback();
+
+  void OnScanCallback(std::shared_ptr<CallbackReply> reply);
+
+  void ProcessScanResult(const std::vector<std::string> &keys);
+
+  void DoParseKeys(const std::vector<std::string> &index_table_keys);
+
+  void DoBatchDelete(const std::vector<std::string> &index_table_keys);
+
+  void DoMultiRead(const std::vector<std::string> &data_table_keys);
+
+  void OnReadCallback(Status status, const boost::optional<std::string> &result,
+                      const std::string &data_key);
+
+  void OnDeleteCallback(Status status);
+
+  std::vector<std::string> DedupeKeys(const std::vector<std::string> &keys);
```

Comment:
`DedupKeys` or `DeduplicateKeys`. "Dedupe" is not the correct abbreviation of "deduplicate".

Reply:
Got it.